db/revision_cache_lru.go

```diff
@@ -274,7 +334,33 @@ func (rc *LRURevisionCache) getValue(docID, revID string, create bool) (value *r
 	} else if create {
 		value = &revCacheValue{key: key}
 		rc.cache[key] = rc.lruList.PushFront(value)
-		for len(rc.cache) > int(rc.capacity) {
+		for rc.lruList.Len() > int(rc.capacity) {
+			rc.purgeOldest_()
+		}
+	}
+	rc.lock.Unlock()
+	return
+}
+
+// getValueByCV gets a value from rev cache by CV, if not found and create is true, will add the value to cache and both lookup maps
+func (rc *LRURevisionCache) getValueByCV(docID string, revID string, cv *CurrentVersionVector, create bool) (value *revCacheValue) {
+	if docID == "" || cv == nil {
+		return nil
+	}
+
+	key := IDandCV{DocID: docID, Source: cv.SourceID, Version: cv.VersionCAS}
+	rc.lock.Lock()
+	if elem := rc.hlvCache[key]; elem != nil {
+		rc.lruList.MoveToFront(elem)
+		value = elem.Value.(*revCacheValue)
+	} else if create {
+		// we need to keep this new element in the CV lookup cache and the revid cache for backwards compatability reasons
+		legacyKey := IDAndRev{DocID: docID, RevID: revID}
```

Comment:
This looks like it will always require callers to know both cv and revID, to ensure that both maps are properly updated.  This wouldn't typically be the case for callers of getFromCache that end up here, I wouldn't think. Let me know if I'm overlooking something, or if you've got a unit test that successfully loads the cache using a cv-based Get, and then fetches from the cache using rev.  (or vice versa)

We probably need to discuss this handling in a bit of depth.

Reply:
I agree makes sense to me. Good spot that was an oversight from me.